File: C/1697.c
#include <stdio.h>  // ¼û¹Ù²ÀÁú
#include <stdlib.h>

#define min(a, b) a > b ? b : a 

typedef struct
{
	struct node* next;
	int data;
}node;

typedef struct
{
	node* rear;
	node* front;
	int count;
}queue;

void init_queue(queue* q)
{
	q->front = q->rear = NULL;
	q->count = 0;
}

int is_empty(queue* q)
{
	return q->count == 0;
}

void push(queue* q, int data)
{
	node* newnode = malloc(sizeof(node));
	newnode->data = data;
	newnode->next = NULL;

	if (is_empty(q))
		q->front = newnode;
	else
		q->rear->next = newnode;
	q->rear = newnode;
	q->count++;
}

int pop(queue* q)
{
	int data;
	node* ptr;
	if (is_empty(q))
		return -1;
	ptr = q->front;
	data = ptr->data;
	q->front = ptr->next;
	free(ptr);
	q->count--;

	return data;
}

int visit[100001] = { 0, };

void bfs(int num, int goal)
{
	queue q;
	init_queue(&q);
	push(&q, num);

	int cnt = 0;
	while (!is_empty(&q))
	{
		int q_size = q.count;
		for (int i = 0; i < q_size; i++)
		{
			int q_pop = pop(&q);
			if (q_pop == goal)
			{
				printf("%d", cnt);
				break;
			}

			if (q_pop - 1 >= 0 && q_pop - 1 <= 100000 && visit[q_pop - 1] == 0)
			{
				visit[q_pop - 1] = 1;
				push(&q, q_pop - 1);
			}
			if (q_pop + 1 >= 0 && q_pop + 1 <= 100000 && visit[q_pop + 1] == 0)
			{
				visit[q_pop + 1] = 1;
				push(&q, q_pop + 1);
			}
			if (q_pop * 2 >= 0 && q_pop * 2 <= 100000 && visit[q_pop * 2] == 0)
			{
				visit[q_pop * 2] = 1;
				push(&q, q_pop * 2);
			}
		}
		cnt++;
	}
}


int main()
{
	int n, k;
	scanf("%d %d", &n, &k);

	bfs(n, k);

	return 0;
}

File: JAVA/src/_1697.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

// 숨바꼭질
public class _1697 {
    static boolean[] visit = new boolean[100001];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int K = Integer.parseInt(input[1]);

        if (N == K) {
            System.out.println(0);
        } else {
            bfs(N, K);
        }
    }

    public static void bfs(int n, int k) {
        Queue<Integer> q = new LinkedList<>();
        q.add(n);
        visit[n] = true;

        int cnt = 0;
        while (!q.isEmpty()) {
            cnt++;
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int num = q.poll();
                visit[num] = true;

                // 위치 찾은 경우
                if (num - 1 == k || num + 1 == k || num * 2 == k) {
                    System.out.println(cnt);
                    return;
                }

                // -1, +1, *2 각 범위별로 체크를 위해 큐에 삽입
                if (num - 1 >= 0 && !visit[num - 1]) {
                    q.add(num - 1);
                    visit[num - 1] = true;
                }
                if (num + 1 <= 100000 && !visit[num + 1]) {
                    q.add(num + 1);
                    visit[num + 1] = true;
                }
                if (num * 2 <= 100000 && !visit[num * 2]) {
                    q.add(num * 2);
                    visit[num * 2] = true;
                }
            }
        }
    }
}
